In [23]:
import pandas as pd
import numpy as np

1. Data Preparation and Overview: Loading the dataset, handling missing values, removing duplicates, and treating outliers.

In [24]:
# Read the CSV file into a DataFrame
df = pd.read_csv('/Users/alex/Downloads/spotify_2020_dataset/spotifytoptracks.csv')


In [25]:
# Set display options for better readability
pd.set_option('display.max_columns', 85)  
pd.set_option('display.max_rows', 85)        
df.head()

Unnamed: 0       artist                                  album  \
0           0   The Weeknd                            After Hours   
1           1  Tones And I                           Dance Monkey   
2           2  Roddy Ricch  Please Excuse Me For Being Antisocial   
3           3    SAINt JHN                  Roses (Imanbek Remix)   
4           4     Dua Lipa                       Future Nostalgia   

              track_name                track_id  energy  danceability  key  \
0        Blinding Lights  0VjIjW4GlUZAMYd2vXMi3b   0.730         0.514    1   
1           Dance Monkey  1rgnBhdG2JDFTbYkYRZAku   0.593         0.825    6   
2                The Box  0nbXyq5TXYPCO7pr3N8S4I   0.586         0.896   10   
3  Roses - Imanbek Remix  2Wo6QQD1KMDWeFkkjLqwx5   0.721         0.785    8   
4        Don't Start Now  3PfIrDoz19wz7qK7tYeu62   0.793         0.793   11   

   loudness  acousticness  speechiness  instrumentalness  liveness  valence  \
0    -5.934       0.00146       0.0598          0.000095    0.0897    0.334   
1    -6.401       0.68800       0.0988          0.000161    0.1700    0.540   
2    -6.687       0.10400       0.0559          0.000000    0.7900    0.642   
3    -5.457       0.01490       0.0506          0.004320    0.2850    0.894   
4    -4.521       0.01230       0.0830          0.000000    0.0951    0.679   

     tempo  duration_ms              genre  
0  171.005       200040           R&B/Soul  
1   98.078       209755  Alternative/Indie  
2  116.971       196653        Hip-Hop/Rap  
3  121.962       176219   Dance/Electronic  
4  123.950       183290           Nu-disco

In [26]:
# Handle missing values
# Check for missing values
missing_values = df.isnull().sum()

print("Missing values in each column:\n", missing_values)

Missing values in each column:
 Unnamed: 0          0
artist              0
album               0
track_name          0
track_id            0
energy              0
danceability        0
key                 0
loudness            0
acousticness        0
speechiness         0
instrumentalness    0
liveness            0
valence             0
tempo               0
duration_ms         0
genre               0
dtype: int64


In [27]:
# Remove unnecessary column
df=df.drop(columns=['Unnamed: 0'])

In [28]:
df.shape

(50, 16)

In [29]:
# Check for duplicate rows
duplicate_rows = df[df.duplicated()]

if not duplicate_rows.empty:
    print(f"There {'is' if len(duplicate_rows) == 1 else 'are'} {len(duplicate_rows)} duplicate row{'' if len(duplicate_rows) == 1 else 's'}.")
else:
    print("There are no duplicate rows.")

There are no duplicate rows.


In [30]:
# Check data types
df.dtypes

artist               object
album                object
track_name           object
track_id             object
energy              float64
danceability        float64
key                   int64
loudness            float64
acousticness        float64
speechiness         float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
duration_ms           int64
genre                object
dtype: object

In [31]:
# Identify numeric columns
numeric_columns = df.select_dtypes(include=['number']).columns

# Convert 'key' to categorical variable if it's a numeric column
if 'key' in numeric_columns:
    df['key'] = df['key'].astype('object')
    print("Converted 'key' to categorical.")
else:
    print("'key' is not a numeric column.")



Converted 'key' to categorical.


In [32]:
# Check data types again
df.dtypes

artist               object
album                object
track_name           object
track_id             object
energy              float64
danceability        float64
key                  object
loudness            float64
acousticness        float64
speechiness         float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
duration_ms           int64
genre                object
dtype: object

Determine outliers in  numerical columns using standard deviation method

In [33]:

# Create a DataFrame to store outliers
outliers_df = pd.DataFrame()
outlier_indices = set()
# Select only numerical columns
numerical_cols = df.select_dtypes(include=['number']).columns

# Define a function to detect outliers using mean and standard deviation method
def detect_outliers(df, column, threshold=3):
    mean = df[column].mean()
    std_dev = df[column].std()
    lower_bound = mean - threshold * std_dev
    upper_bound = mean + threshold * std_dev
    return df[(df[column] < lower_bound) | (df[column] > upper_bound)]

# Apply the outlier detection for each numerical column
for col in numerical_cols:
    outliers = detect_outliers(df, col)
    outliers_df = pd.concat([outliers_df, outliers])
    outlier_indices.update(outliers.index)

# Drop duplicate outliers (if any)
outliers_df = outliers_df.drop_duplicates()

# Convert outlier indices to a sorted list
outlier_indices = sorted(outlier_indices)

print("Outlier indices:", outlier_indices)

outliers_df.head()

Outlier indices: [2, 19, 24, 41, 49]


artist                                  album  \
24    Billie Eilish                    everything i wanted   
19           Future                          High Off Life   
2       Roddy Ricch  Please Excuse Me For Being Antisocial   
41  Black Eyed Peas                            Translation   
49     Travis Scott                             ASTROWORLD   

                    track_name                track_id  energy  danceability  \
24         everything i wanted  3ZCTVFBt2Brf31RLEnCkWJ   0.225         0.704   
19  Life Is Good (feat. Drake)  1K5KBOgreBi5fkEHvg5ap3   0.574         0.795   
2                      The Box  0nbXyq5TXYPCO7pr3N8S4I   0.586         0.896   
41   RITMO (Bad Boys For Life)  4NCsrTzgVfsDo8nWyP8PPc   0.704         0.723   
49                  SICKO MODE  2xLMifQCjDGFmkHkpNLD9h   0.730         0.834   

   key  loudness  acousticness  speechiness  instrumentalness  liveness  \
24   6   -14.454       0.90200       0.0994           0.65700     0.106   
19   2    -6.903       0.06700       0.4870           0.00000     0.150   
2   10    -6.687       0.10400       0.0559           0.00000     0.790   
41  10    -7.088       0.02590       0.0571           0.00109     0.792   
49   8    -3.714       0.00513       0.2220           0.00000     0.124   

    valence    tempo  duration_ms        genre  
24    0.243  120.006       245426          Pop  
19    0.537  142.053       237918  Hip-Hop/Rap  
2     0.642  116.971       196653  Hip-Hop/Rap  
41    0.684  105.095       214935          Pop  
49    0.446  155.008       312820  Hip-Hop/Rap

For the analysis below the dataframe has been used with outliers

In [34]:

# Number of observations
num_observations = df.shape[0]
print("Number of observations:", num_observations)


Number of observations: 50


In [35]:
# Number of features
num_features = df.shape[1]
print("Number of features:", num_features)

Number of features: 16


In [36]:
# Categorical features
categorical_features = df.select_dtypes(include=['object']).columns
print("Categorical features:", categorical_features)

Categorical features: Index(['artist', 'album', 'track_name', 'track_id', 'key', 'genre'], dtype='object')


In [37]:
# Numeric features
numeric_features = df.select_dtypes(include=['number']).columns
print("Numeric features:", numeric_features)

Numeric features: Index(['energy', 'danceability', 'loudness', 'acousticness', 'speechiness',
       'instrumentalness', 'liveness', 'valence', 'tempo', 'duration_ms'],
      dtype='object')


2. Artist and Track Analysis: Analysis of popular artists, albums, track characteristics, and genres.

In [38]:
# Artists with more than 1 popular track
artist_track_counts = df['artist'].value_counts()
artists_with_multiple_tracks = artist_track_counts[artist_track_counts > 1]
print("Artists with more than 1 popular track:\n", artists_with_multiple_tracks)

Artists with more than 1 popular track:
 artist
Dua Lipa         3
Billie Eilish    3
Travis Scott     3
Harry Styles     2
Lewis Capaldi    2
Justin Bieber    2
Post Malone      2
Name: count, dtype: int64


In [39]:
# The most popular artist

# Find the maximum count of tracks
max_count = artist_track_counts.max()

# Find all artists with the maximum count of tracks
most_popular_artists = artist_track_counts[artist_track_counts == max_count]

print(f"The most popular artist(s) with {max_count} tracks:\n{most_popular_artists}")

The most popular artist(s) with 3 tracks:
artist
Dua Lipa         3
Billie Eilish    3
Travis Scott     3
Name: count, dtype: int64


In [40]:
# Total number of  artists in the top 50

# Count unique artists
num_unique_artists = len(df['artist'].unique())

print(f'Total number of unique artists : {num_unique_artists} artists')

Total number of unique artists : 40 artists


In [41]:
# Albums that have more than 1 popular track

# Group by the 'album' column and count the number of tracks per album
album_track_counts = df.groupby('album')['track_name'].count()

# Filter albums with more than one popular track
albums_with_multiple_tracks = album_track_counts[album_track_counts > 1]

# Sort the albums by the number of popular tracks in descending order
albums_with_multiple_tracks = albums_with_multiple_tracks.sort_values(ascending=False)

# Display the albums and the number of popular tracks they have
print(albums_with_multiple_tracks)


album
Future Nostalgia        3
Changes                 2
Fine Line               2
Hollywood's Bleeding    2
Name: track_name, dtype: int64


In [42]:
# Number of albums in top 50

# Extract the unique album names and count them
num_unique_albums = len(df['album'].unique())

print(f'Total number of unique albums in the top 50: {num_unique_albums}')

Total number of unique albums in the top 50: 45


In [43]:
# Tracks that have a danceability score above 0.7

# Filter tracks 
high_danceability_tracks = df[df['danceability'] > 0.7]
sorted_high_danceability_tracks = high_danceability_tracks.sort_values(by='danceability', ascending=False)

# Count the number of tracks with danceability score above 0.7
num_high_danceability_tracks = high_danceability_tracks.shape[0]

print(f"Number of tracks with danceability score above 0.7: {num_high_danceability_tracks}")

print("Tracks with danceability score above 0.7:")
print(sorted_high_danceability_tracks[['track_name', 'danceability']])



Number of tracks with danceability score above 0.7: 32
Tracks with danceability score above 0.7:
                                       track_name  danceability
27                WAP (feat. Megan Thee Stallion)         0.935
2                                         The Box         0.896
39                                        Ride It         0.880
28                                    Sunday Best         0.878
33               Supalonely (feat. Gus Dapperton)         0.862
40                                     goosebumps         0.841
49                                     SICKO MODE         0.834
15                                   Toosie Slide         0.830
1                                    Dance Monkey         0.825
29                    Godzilla (feat. Juice WRLD)         0.808
14                       Intentions (feat. Quavo)         0.806
10                                           Tusa         0.803
19                     Life Is Good (feat. Drake)         0.795
4      

In [44]:
# Tracks with danceability score below 0.4

# Filter tracks 
low_danceability_tracks = df[df['danceability'] < 0.4]

print("\nTracks with danceability score below 0.4:")
print(low_danceability_tracks[['track_name', 'danceability']])


Tracks with danceability score below 0.4:
              track_name  danceability
44  lovely (with Khalid)         0.351


In [45]:
# Tracks with loudness above -5

# Filter tracks with loudness above -5
high_loudness_tracks = df[df['loudness'] > -5]
sorted_high_loudness_tracks = high_loudness_tracks.sort_values(by='loudness', ascending=False)

# Count the number of tracks with loudness above -5
num_hloudness = high_loudness_tracks.shape[0]
print(f"Number of tracks with loudness above -5: {num_hloudness}")

print("Tracks with loudness above -5:")
print(sorted_high_loudness_tracks[['track_name', 'loudness']])


Number of tracks with loudness above -5: 19
Tracks with loudness above -5:
                                       track_name  loudness
10                                           Tusa    -3.280
40                                     goosebumps    -3.370
31                                 Break My Heart    -3.434
38                                          Hawái    -3.454
12                                        Circles    -3.497
23                         Mood (feat. iann dior)    -3.558
21                                      Adore You    -3.675
49                                     SICKO MODE    -3.714
48                                       Physical    -3.756
35                Rain On Me (with Ariana Grande)    -3.764
43                                        Safaera    -4.074
6                                Watermelon Sugar    -4.209
39                                        Ride It    -4.258
37  Sunflower - Spider-Man: Into the Spider-Verse    -4.368
32                       

In [46]:
# Tracks with loudness below -8

# Filter tracks with loudness below -8
low_loudness_tracks = df[df['loudness'] < -8]
sorted_low_loudness_tracks = low_loudness_tracks.sort_values(by='loudness', ascending=False)
# Count the number of tracks with loudness below -8
num_lloudness = low_loudness_tracks.shape[0]
print(f"Number of tracks with loudness below -8: {num_lloudness}")

print("\nTracks with loudness below -8:")
print(sorted_low_loudness_tracks[['track_name', 'loudness']])

Number of tracks with loudness below -8: 9

Tracks with loudness below -8:
                                        track_name  loudness
20                Savage Love (Laxed - Siren Beat)    -8.520
8                                          Falling    -8.756
36                             HIGHEST IN THE ROOM    -8.764
7                 death bed (coffee for your head)    -8.765
15                                    Toosie Slide    -8.820
47  If the World Was Ending - feat. Julia Michaels   -10.086
44                            lovely (with Khalid)   -10.109
26                                         bad guy   -10.965
24                             everything i wanted   -14.454


In [47]:
# The longest track

# Find the row with the maximum duration
longest_track = df.loc[df['duration_ms'].idxmax()]

print(longest_track[['artist', 'track_name', 'duration_ms']])

artist         Travis Scott
track_name       SICKO MODE
duration_ms          312820
Name: 49, dtype: object


In [48]:
# The shortest track

# Find the row with the min duration
shortest_track = df.loc[df['duration_ms'].idxmin()]

print(shortest_track[['artist', 'track_name', 'duration_ms']])

artist                       24kGoldn
track_name     Mood (feat. iann dior)
duration_ms                    140526
Name: 23, dtype: object


In [49]:
# The most popular genre

# Group by 'genre' and count the number of tracks in each genre
genre_counts = df['genre'].value_counts()

# Sort the genres by their count
sorted_genre_counts = genre_counts.sort_values(ascending=False)

# Get the most popular genre
most_popular_genre = genre_counts.idxmax()
most_popular_count = genre_counts.max()

print("Sorted genre counts:")
print(sorted_genre_counts)
print(f"\nThe most popular genre is '{most_popular_genre}' with {most_popular_count} tracks.")

Sorted genre counts:
genre
Pop                                   14
Hip-Hop/Rap                           13
Dance/Electronic                       5
Alternative/Indie                      4
R&B/Soul                               2
 Electro-pop                           2
R&B/Hip-Hop alternative                1
Nu-disco                               1
Pop/Soft Rock                          1
Pop rap                                1
Hip-Hop/Trap                           1
Dance-pop/Disco                        1
Disco-pop                              1
Dreampop/Hip-Hop/R&B                   1
Alternative/reggaeton/experimental     1
Chamber pop                            1
Name: count, dtype: int64

The most popular genre is 'Pop' with 14 tracks.


In [50]:
# Genres that have just one song on the top 50

# Filter genres with only one track
genres_with_one_song = genre_counts[genre_counts == 1]

print("Genres with just one song in the top 50:")
print(genres_with_one_song)

Genres with just one song in the top 50:
genre
R&B/Hip-Hop alternative               1
Nu-disco                              1
Pop/Soft Rock                         1
Pop rap                               1
Hip-Hop/Trap                          1
Dance-pop/Disco                       1
Disco-pop                             1
Dreampop/Hip-Hop/R&B                  1
Alternative/reggaeton/experimental    1
Chamber pop                           1
Name: count, dtype: int64


In [51]:
# Number of genres in top 50

# Count the number of unique genres
total_genres = df['genre'].nunique()

print(f"Total number of genres represented in the top 50: {total_genres}")


Total number of genres represented in the top 50: 16


3. Feature Relationships and Comparisons:** Correlation analysis and genre comparisons.

Correlation analysis
Let's specify correlation coefficents for current research:
Strong positive correlation R > 0.6
Strong negative correlation R < - 0.6
No correlation -0.1 < R < 0.1

In [52]:

# Select only numerical features
numerical_df = df.select_dtypes(include=['number'])

numerical_df.head()

energy  danceability  loudness  acousticness  speechiness  \
0   0.730         0.514    -5.934       0.00146       0.0598   
1   0.593         0.825    -6.401       0.68800       0.0988   
2   0.586         0.896    -6.687       0.10400       0.0559   
3   0.721         0.785    -5.457       0.01490       0.0506   
4   0.793         0.793    -4.521       0.01230       0.0830   

   instrumentalness  liveness  valence    tempo  duration_ms  
0          0.000095    0.0897    0.334  171.005       200040  
1          0.000161    0.1700    0.540   98.078       209755  
2          0.000000    0.7900    0.642  116.971       196653  
3          0.004320    0.2850    0.894  121.962       176219  
4          0.000000    0.0951    0.679  123.950       183290

In [53]:
# Calculate the correlation matrix for numerical features
correlation_matrix = numerical_df.corr()

correlation_matrix

energy  danceability  loudness  acousticness  speechiness  \
energy            1.000000      0.152552  0.791640     -0.682479     0.074267   
danceability      0.152552      1.000000  0.167147     -0.359135     0.226148   
loudness          0.791640      0.167147  1.000000     -0.498695    -0.021693   
acousticness     -0.682479     -0.359135 -0.498695      1.000000    -0.135392   
speechiness       0.074267      0.226148 -0.021693     -0.135392     1.000000   
instrumentalness -0.385515     -0.017706 -0.553735      0.352184     0.028948   
liveness          0.069487     -0.006648 -0.069939     -0.128384    -0.142957   
valence           0.393453      0.479953  0.406772     -0.243192     0.053867   
tempo             0.075191      0.168956  0.102097     -0.241119     0.215504   
duration_ms       0.081971     -0.033763  0.064130     -0.010988     0.366976   

                  instrumentalness  liveness   valence     tempo  duration_ms  
energy                   -0.385515  0.069487  0.393453  0.075191     0.081971  
danceability             -0.017706 -0.006648  0.479953  0.168956    -0.033763  
loudness                 -0.553735 -0.069939  0.406772  0.102097     0.064130  
acousticness              0.352184 -0.128384 -0.243192 -0.241119    -0.010988  
speechiness               0.028948 -0.142957  0.053867  0.215504     0.366976  
instrumentalness          1.000000 -0.087034 -0.203283  0.018853     0.184709  
liveness                 -0.087034  1.000000 -0.033366  0.025457    -0.090188  
valence                  -0.203283 -0.033366  1.000000  0.045089    -0.039794  
tempo                     0.018853  0.025457  0.045089  1.000000     0.130328  
duration_ms               0.184709 -0.090188 -0.039794  0.130328     1.000000

In [54]:
# Strongly positively correlated features
# Find pairs of features with strong positive correlation (correlation > 0.6)
strong_positive_corr = correlation_matrix[(correlation_matrix > 0.6) & (correlation_matrix < 1.0)]

# Drop NaN values to focus on strong correlations
strong_positive_corr = strong_positive_corr.dropna(how='all').dropna(axis=1, how='all')

print("Strongly positively correlated features:\n", strong_positive_corr)

Strongly positively correlated features:
            energy  loudness
energy        NaN   0.79164
loudness  0.79164       NaN


Energy and loudness are strongly positively correlated.

In [55]:
# Strongly negatively correlated features
# Find pairs of features with strong negative correlation (correlation < -0.6)
strong_negative_corr = correlation_matrix[correlation_matrix < -0.6]

# Drop NaN values to focus on strong correlations
strong_negative_corr = strong_negative_corr.dropna(how='all').dropna(axis=1, how='all')

print("Strongly negatively correlated features:\n", strong_negative_corr)

Strongly negatively correlated features:
                 energy  acousticness
energy             NaN     -0.682479
acousticness -0.682479           NaN


Energy and lacousticness are strongly negatively correlated.

In [56]:
# Features that are not correlated

# Find pairs of features with weak correlation (correlation between -0.1 and 0.1)
not_correlated = correlation_matrix[(correlation_matrix > -0.1) & (correlation_matrix < 0.1)]

# Drop NaN values to focus on weak correlations
not_correlated = not_correlated.dropna(how='all').dropna(axis=1, how='all')

print("Features that are not correlated:\n", not_correlated)


Features that are not correlated:
                     energy  danceability  loudness  acousticness  speechiness  \
energy                 NaN           NaN       NaN           NaN     0.074267   
danceability           NaN           NaN       NaN           NaN          NaN   
loudness               NaN           NaN       NaN           NaN    -0.021693   
acousticness           NaN           NaN       NaN           NaN          NaN   
speechiness       0.074267           NaN -0.021693           NaN          NaN   
instrumentalness       NaN     -0.017706       NaN           NaN     0.028948   
liveness          0.069487     -0.006648 -0.069939           NaN          NaN   
valence                NaN           NaN       NaN           NaN     0.053867   
tempo             0.075191           NaN       NaN           NaN          NaN   
duration_ms       0.081971     -0.033763  0.064130     -0.010988          NaN   

                  instrumentalness  liveness   valence     tempo  duratio

A lot of the features in the dataset don't really affect each other. For example, there's almost no connection between acousticness and duration, or between valence and liveness. Acousticness only has one feature that it doesn't correlate with, while liveness doesn't correlate with seven other features.

In [57]:
# Comparison of the danceability, loudness and acousticness scores between Pop, 
# Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie genres

# Define selected features and genres
selected_features = ['danceability', 'loudness', 'acousticness']
selected_genres = ['Pop', 'Hip-Hop/Rap', 'Dance/Electronic', 'Alternative/Indie']

# Perform groupby and aggregate mean, max, and min
genre_stats = df[df['genre'].isin(selected_genres)].groupby('genre')[selected_features].agg(['mean', 'max', 'min'])

print(genre_stats)

                  danceability                loudness                 \
                          mean    max    min      mean    max     min   
genre                                                                   
Alternative/Indie     0.661750  0.862  0.459 -5.421000 -4.746  -6.401   
Dance/Electronic      0.755000  0.880  0.647 -5.338000 -3.756  -7.567   
Hip-Hop/Rap           0.765538  0.896  0.598 -6.917846 -3.370  -8.820   
Pop                   0.677571  0.806  0.464 -6.460357 -3.280 -14.454   

                  acousticness                  
                          mean    max      min  
genre                                           
Alternative/Indie     0.583500  0.751  0.29100  
Dance/Electronic      0.099440  0.223  0.01370  
Hip-Hop/Rap           0.188741  0.731  0.00513  
Pop                   0.323843  0.902  0.02100  


Danceability. Hip-Hop/Rap and Dance/Electronic genres have the highest mean danceability value (0.755-0.766).
These genres are generally very suitable for dancing. Alternative/Indie and Pop have similar danceability mean scores (0.662-0.678). Alternative/Indie has the widest range od danceability.

Loudness. The quietest genre  is Hip-Hop / Rap with -6.92 mean value. Pop songs have the widest range of loudness, suggesting that some songs can be very loud while others are quite soft(max= -3.280, min= -14.454 ).  On average Dance/Electronic songs are the loudest(-5.34).

Acousticness. The most acoustic genre is Alternative / Indie with 0.58 mean value. Dance/Electronic and Hip-Hop/Rap are less acoustic on average. Pop songs could be as very much and not acoustic at all(min=0.02, max=0.9).

Areas of improvement for further research.

1. Increase the number of observations in the dataset
2. Consider usage of visualization
3. Check statistical signifigance of correlation coefficients
4. Compare features of popular with less popular tracks to identify what makes the song popular?
5. Include user engagement metrics in the dataset

